### PR TITLE
Universus warper

### PR DESCRIPTION
**Pull request by stanis-chernes**

### CODE REVIEW

File: contracts/universus/UniversusWarper.sol [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    function getLastActiveRentalId(address renter, uint256 tokenId) external view returns (uint256) {
+        return _lastActiveRental[renter][tokenId];
+    }
+
+    /**
+     * @dev Returns RentalDetails for 'rentalId'.
+     * @param rentId Rent ID.
+     * @return RentalDetails.
+     */
+    function getRentalDetails(uint256 rentId) public view returns (IUniversusWarper.RentalDetails memory) {
```
Comment (stanis-chernes) [inline contracts/universus/UniversusWarper.sol:53]: correct rentId -> rentalID

File: contracts/universus/UniversusWarper.sol [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            _lastActiveRental[rentalAgreement.renter][tokenId] = rentalId;
+
+            RentalDetails storage rentalDetails = _rentalDetails[rentalId];
+            rentalDetails.listingTerms = rentalAgreement.agreementTerms.listingTerms;
+            rentalDetails.universeTaxTerms = rentalAgreement.agreementTerms.universeTaxTerms;
+            rentalDetails.protocolTaxTerms = rentalAgreement.agreementTerms.protocolTaxTerms;
+            rentalDetails.rentalId = rentalId;
+            rentalDetails.listingId = rentalAgreement.listingId;
+        }
+
```
Comment (stanis-chernes) [inline contracts/universus/UniversusWarper.sol:92]: add event OnRentHookEvent(
      address renter,
      uint256 tokenId,
      uint256 rentalId
)

File: contracts/universus/IUniversusWarper.sol [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+     * @param universeTaxTerms Universe tax terms information.
+     * @param protocolTaxTerms Protocol tax terms information.
+     * @param rentalId Asset rental ID.
+     * @param listingId Asset listing ID.
+     */
+    struct RentalDetails {
+        IListingTermsRegistry.ListingTerms listingTerms;
+        ITaxTermsRegistry.TaxTerms universeTaxTerms;
+        ITaxTermsRegistry.TaxTerms protocolTaxTerms;
+        uint256 rentalId;
```
Comment (stanis-chernes) [inline contracts/universus/IUniversusWarper.sol:20]: address lister,
address protocol

File: contracts/universus/IUniversusWarper.sol [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    /**
+     * @dev Struct defining the rental details.
+     * @param listingTerms Listing terms information.
+     * @param universeTaxTerms Universe tax terms information.
+     * @param protocolTaxTerms Protocol tax terms information.
+     * @param rentalId Asset rental ID.
+     * @param listingId Asset listing ID.
+     */
+    struct RentalDetails {
+        IListingTermsRegistry.ListingTerms listingTerms;
```
Comment (stanis-chernes) [inline contracts/universus/IUniversusWarper.sol:17]: add method getUniverseRewardAddress()

File: contracts/universus/UniversusWarper.sol [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+     * @return RentalDetails.
+     */
+    function getRentalDetails(uint256 rentId) public view returns (IUniversusWarper.RentalDetails memory) {
+        return _rentalDetails[rentId];
+    }
+
+    /**
+     * @dev Constructor for the IQNFTWarper contract.
+     */
+    constructor(bytes memory config) Auth() warperInitializer {
```
Comment (stanis-chernes) [inline contracts/universus/UniversusWarper.sol:60]: add here universeRewardAddress

File: contracts/universus/UniversusWarper.sol [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        super.__initialize(config);
+
+        // (, , address rewardPool) = abi.decode(config, (address, address, address));
```
Comment (stanis-chernes) [inline contracts/universus/UniversusWarper.sol:63]: take universeRewardAddress from abi.decode of bytes memory config

File: contracts/universus/UniversusWarper.sol [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+/**
+ * @title Custom Warper for Universus
+ */
+contract UniversusWarper is IUniversusWarper, IRentingHookMechanics, ERC721ConfigurablePreset, Auth {
+    /**
+     * @dev ListingManager contact key.
+     */
+    bytes4 private immutable LISTING_MANAGER;
+
+    /**
```
Comment (stanis-chernes) [inline contracts/universus/UniversusWarper.sol:23]: address private universeRewardAddress

File: contracts/universus/UniversusWarper.sol (hunk 6 above)
Comment (stanis-chernes) [reply contracts/universus/UniversusWarper.sol:63]: universeRewardAddress = ...

File: contracts/universus/UniversusWarper.sol [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+            (, uint256 tokenId) = _decodeAssetId(rentalAgreement.warpedAssets[i].id);
+            // Latest active rental is persisted.
+            _lastActiveRental[rentalAgreement.renter][tokenId] = rentalId;
+
+            RentalDetails storage rentalDetails = _rentalDetails[rentalId];
+            rentalDetails.listingTerms = rentalAgreement.agreementTerms.listingTerms;
+            rentalDetails.universeTaxTerms = rentalAgreement.agreementTerms.universeTaxTerms;
+            rentalDetails.protocolTaxTerms = rentalAgreement.agreementTerms.protocolTaxTerms;
+            rentalDetails.rentalId = rentalId;
+            rentalDetails.listingId = rentalAgreement.listingId;
```
Comment (stanis-chernes) [inline contracts/universus/UniversusWarper.sol:90]: rentalDetails.lister = ListingManager(...).listingInfo(rentalAgreement.listing).beneficiary;

Comment (stanis-chernes) [reply contracts/universus/UniversusWarper.sol:90]: rentalDetails.protocol =  IMetahub(_metahub()).protocolExternalFeesCollector();

File: contracts/universus/UniversusWarper.sol (hunk 5 above)
Comment (Alenorze) [reply contracts/universus/UniversusWarper.sol:46]: Okay, no problem

**Pull request merged**